src/main/resources/graphql/schema.graphqls

```diff
@@ -1,16 +1,110 @@
-# schema file of the microservice
+type Query {
+"""
+Get the skill levels of the current user for all skill types (Bloom taxonomy) in a chapter of a course.
+If chapterId is null, then the skill levels for the course are returned.
+"""
+userSkillLevels(courseId: UUID!, chapterId: UUID): SkillLevels!
 
-# TODO define data types, queries and mutations
-# this can be done in a separate files as long as they are in this folder and
-# end with .graphqls
+"""
+Get the skill levels of the specified user for all skill types (Bloom taxonomy) in a chapter of a course.
+If chapterId is null, then the skill levels for the course are returned.
+"""
+skillLevelsForUser(userId: UUID!, courseId: UUID!, chapterId: UUID): SkillLevels!
+}
 
+"""
+The four skill level of a user.
+"""
+type SkillLevels {
+  """
+  remember represents how much user remember the concept
+  """
+  remember: SkillLevel!
+  """
+  understand represents how well the user understands learned content.
+  """
+  understand: SkillLevel!
 
-# example
-type Template {
-  id: ID!
-  name: String!
+  """
+  apply represents the how well user applies the learned concept during assessment.
+  """
+  apply: SkillLevel!
+  """
+  apply is how much user can evaluate information and draw conclusions
+  """
+  analyze: SkillLevel!
+}
+"""
+The skill level of a user.
+"""
+type SkillLevel {
+  """
+  The value of the skill level.
+  levels are between 0 and 100.
+  """
+  value: Int!
+  """
+  A log of the changes to the skill level
+  """
+  log: [SkillLevelLogItem!]!
 }
 
-type Query {
-  templates: [Template]
-}
\ No newline at end of file
+"""
+An item in the skill level change log.
+"""
+type SkillLevelLogItem {
+  """
+  The date when the skill level changed.
+  """
+  date: DateTime!
+  """
+  The difference between the previous and the new skill level.
+  """
+  difference: Int!
+  """
+  The old skill level.
+  """
+  oldValue: Int!
+  """
+  The new skill level.
+  """
+  newValue: Int!
+  """
+  The reason why the skill level has changed.
+  """
+  reason: SkillLevelChangeReason!
+  """
+  The ids of the chapters that are associated with the change.
+  """
+  associatedChapterIds: [UUID!]!
+}
+"""
+The reason why the skill level has changed.
+"""
+enum SkillLevelChangeReason {
+  """
+  The user failed in an assessment, resulting in a decrease in skill levels.
+  """
+  ASSESSMENT_FAILED,
+  """
+  The user learned from a new assessment or repeated an already done assessment,
+  resulting in an increase in skill levels.
+  """
+  ASSESSMENT_LEARNING,
+  """
+  The user repeated an already done assessment, resulting in an increase in skill levels.
+  """
+  ASSESSMENT_REPEATING,
+  """
+  The user used hints while attempting an assessment, resulting in a smaller increase in skill levels.
+  """
+  ASSESSMENT_WITH_HINTS,
```

Comment:
That is an interesting idea! But what would happen if both hints were used and the correctness changes?

Reply:
By using a weighted approach, we can maintain a fair representation of the user's actual skill level while taking into account the influence of hints and the correctness of responses. 
Weighting the Factors: To handle the combination of hints used and correctness changes, we can apply weighting to each factor. For example, we could assign percentages to each factor to determine their contribution to the final skill level change.
scenario could be: 
1. If the user uses hints and responds correctly, the skill level increase might be, for example, 30% of what it would be without hints.
2. If the user uses hints and responds incorrectly, the skill level might increase by only 10% of what it would be without hints.
3. If the user doesn't use hints and responds correctly, the skill level would increase significantly, let's say 90% of what it would be without hints.
4. If the user doesn't use hints and responds incorrectly, the skill level might not increase or even decrease slightly.

is it good idea? what you think @PaulBredl

Reply:
because I created that reason to looked into concept that we have for skill level: 
Skill levels are increased... 
• when learning new assessments or repeating
already done assessments
• less if student used hints 
• less depending on
correctness 
• depending on the time (on
timed assessments)